amandaguan-ag/todo
# #2 task prioritization

## Type of Changes



|     | Type                       |
| --- | -------------------------- |
|  ✓    | :sparkles: New feature     |
|  ✓    | :link: Update dependencies |

## Acceptance Criteria

- [x] Update the database schema to include a priority field for tasks.
- [x] Modify the task addition form to include priority selection (e.g., Low, Medium, High).
- [x] Update API endpoints to handle task priority.
- [x] Implement logic for sorting tasks based on priority in the UI.
- [x] Implement logic for sorting tasks by creation date when priorities are the same.

## Related Issue

Closes #4 

## Description 

This PR introduces significant enhancements to the task management capabilities of my application, focusing on task display and sorting based on priority and creation date. These updates aim to provide users with an intuitive task management experience, allowing for easier organization and prioritization of their tasks.

### Key Features and Enhancements

- **Enhanced Task Sorting**: Tasks are now sorted primarily by priority. For tasks with the same priority level, I introduced secondary sorting by their creation date, from oldest to newest. This ensures tasks are organized in a logical order that reflects both urgency and timeliness.

- **UI Improvements for Task Display**: To improve clarity and aesthetics, I've refined the task display. Priority levels are now seamlessly integrated into the task description without brackets, and creation dates are formatted in a user-friendly manner.

### Testing Steps / QA Criteria

To prepare for testing, please ensure you've followed the [setup steps](https://github.com/amandaguan-ag/todo#setup) outlined in the project's README, including running any necessary database migrations to update your local database schema.

1. Checkout to this branch with `git checkout 2-task-prioritization`.
2. Make sure you are in the backend directory and run the database migrations to update your schema by executing `npm run typeorm migration:run`.
3. Start the backend service by running `npm run start` in the backend directory.
4. In a separate terminal, navigate to the frontend directory and run `npm start` to launch the frontend.
5. Navigate to [http://localhost:3000](http://localhost:3000) to access the Todo List Application.
6. Create tasks with varying priorities (Low, Medium, High) and observe the sorting order.
7. Verify that tasks with identical priorities are sorted by their creation date, with older tasks appearing first in the list.

## Updates

### Before


<img width="1840" alt="image" src="https://github.com/amandaguan-ag/todo/assets/121919044/d60cee96-c29d-406d-978d-90f6abe00263">

### After


<img width="1842" alt="image" src="https://github.com/amandaguan-ag/todo/assets/121919044/6f8a76a3-1e27-4c8f-9768-5ad727fac141">
<img width="556" alt="image" src="https://github.com/amandaguan-ag/todo/assets/121919044/93b6e4db-a996-4edb-b208-e14cf6e235d5">


## TODOs

- Add unit tests for the new sorting functionality.
- Implement task drag and drop functionality using react-beautiful-dnd for a more intuitive and user-friendly task management experience.
